### ISSUE TITLE
Bug in optimiser(maybe)

### ISSUE THREAD

**Issue by eklavya** [rust-lang/rust#22008]
Hi,

The rust optimiser or some other component is not behaving correctly resulting in wrong results.
I have produced a small example here : https://github.com/eklavya/rust-bug

This was reproduced on 
Mac yosemite:
cargo 0.0.1-pre-nightly (2b9a3d6 2015-01-30 02:47:30 +0000)
rustc 1.0.0-nightly (3b2ed1490 2015-02-03 14:56:32 +0000)
and 
centos7 x64, rustc 1.0.0-nightly (706be5ba1 2015-02-05 23:14:28 +0000)

Please let me know if you need any more details.

**Comment by renato-zannon** [comment]
Reduced test case:
```
fn main() {
    let command = "a";

    match command {
        "foo" => println!("foo"),
        _     => println!("{}", command),
    }
}
```

This segfaults with `opt-level` >= 2.

**Comment by renato-zannon** [comment]
This goes back to at least 0.12 - though it prints garbage values instead of segfaulting. 0.10 seems unaffected (I couldn't find a 0.11 archive for linux).

**Comment by eddyb** [comment]
I've some some hacky [introspection](https://play.rust-lang.org/?code=%23%5Binline%28never%29%5D%0Afn%20format%28args%3A%20std%3A%3Afmt%3A%3AArguments%29%20%7B%0A%20%20%20%20let%20args%20%3D%20unsafe%20%7B%0A%20%20%20%20%20%20%20%20std%3A%3Amem%3A%3Atransmute%3A%3A%3C_%2C%20%28%26%5Bu8%5D%2C%20Option%3C%26%5Bu8%5D%3E%2C%20%26%5B%28%26%26%5Bu8%5D%2C%20fn%28%29%29%5D%29%3E%28args%29%0A%20%20%20%20%7D%3B%0A%20%20%20%20let%20arg%20%3D%20args.2%5B0%5D%3B%0A%20%20%20%20let%20overwritten%20%3D%20arg.0.len%28%29%20as%20*const%20%5Bu8%3B%2032%5D%3B%0A%20%20%20%20println%21%28%22%7B%3Ap%7D%20%7B%3A%3F%7D%22%2C%20overwritten%2C%20unsafe%20%7B*overwritten%7D%29%3B%0A%7D%0A%0Afn%20main%28%29%20%7B%0A%20%20%20%20let%20command%20%3D%20b%22a%22%3B%0A%0A%20%20%20%20match%20command%20%7B%0A%20%20%20%20%20%20%20%20b%22%22%20%3D%3E%20%7B%7D%0A%20%20%20%20%20%20%20%20_%20%3D%3E%20format%28format_args%21%28%22%7B%3A%3F%7D%22%2C%20command%29%29%0A%20%20%20%20%7D%0A%7D).

**Comment by dotdash** [comment]
This has two parts. When I extended rustc to emit lifetime intrinsics, I chose to put a call to lifetime.end at the end of functions to mark arguments as dead that are passed by value on the stack. This allows LLVM to remove dead stores at the end of functions that it normally can't remove. This works because normally, the caller makes a copy of the value it passes in.

Now, the call that match generates to compare the strings is to `eq_slice`, which accepts two slices by-value, and slices are large enough to be passed on the stack. Thus, calls to lifetime.end are emitted. But, the call the match generates does *not* actually make copies of the slices, but passes in pointers to the original slices. That means that the lifetime.end call marks the original alloca as dead, and then things fall apart.

Question: Do we want to drop the lifetime.end call for by-value on-stack arguments, or fix the call to eq_slice to make copies?

cc @nikomatsakis @brson
